generators/tracegen/src/main/scala/Tile.scala

```diff
@@ -29,6 +34,190 @@ class TraceGenTile(val id: Int, val params: TraceGenParams, q: Parameters)
   override lazy val module = new TraceGenTileModuleImp(this)
 }
 
+class BoomLSUShim(implicit p: Parameters) extends BoomModule()(p)
+  with MemoryOpConstants {
+  val io = IO(new Bundle {
+    val lsu = Flipped(new LSUCoreIO)
+    val tracegen = Flipped(new HellaCacheIO)
+  })
+
+  io.lsu.tsc_reg := 0.U(1.W)
+
+  val rob_sz = numRobEntries
+  val rob = Reg(Vec(rob_sz, new HellaCacheReq))
+  val rob_respd = RegInit(VecInit((~(0.U(rob_sz.W))).asBools))
+  val rob_uop = Reg(Vec(rob_sz, new MicroOp))
+  val rob_bsy  = RegInit(VecInit(0.U(rob_sz.W).asBools))
+  val rob_head = RegInit(0.U(log2Up(rob_sz).W))
+  val rob_tail = RegInit(0.U(log2Up(rob_sz).W))
+  val rob_wait_till_empty = RegInit(false.B)
+  val ready_for_amo = rob_tail === rob_head && io.lsu.fencei_rdy
+  when (ready_for_amo) {
+    rob_wait_till_empty := false.B
+  }
+
+  def WrapInc(idx: UInt, max: Int): UInt = {
+    Mux(idx === (max-1).U, 0.U, idx + 1.U)
+  }
+
+
+  io.tracegen.req.ready := (!rob_bsy(rob_tail) &&
+    !rob_wait_till_empty &&
+    (ready_for_amo || !(isAMO(io.tracegen.req.bits.cmd) || io.tracegen.req.bits.cmd === M_XLR || io.tracegen.req.bits.cmd === M_XSC)) &&
+    (WrapInc(rob_tail, rob_sz) =/= rob_head) &&
```

Comment:
This always leaves one entry in the ROB empty. Take a look at Chisel's queue implementation to see how you can detect full queues properly.